phoenix-tools/src/main/java/org/apache/phoenix/schema/SchemaExtractionProcessor.java

```diff
@@ -0,0 +1,410 @@
+package org.apache.phoenix.schema;
+
+import com.google.common.collect.Sets;
+import org.apache.commons.lang.StringUtils;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.HColumnDescriptor;
+import org.apache.hadoop.hbase.HTableDescriptor;
+import org.apache.hadoop.hbase.TableName;
+import org.apache.hadoop.hbase.io.ImmutableBytesWritable;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.apache.phoenix.jdbc.PhoenixConnection;
+import org.apache.phoenix.mapreduce.util.ConnectionUtil;
+import org.apache.phoenix.query.ConnectionQueryServices;
+import org.apache.phoenix.util.MetaDataUtil;
+import org.apache.phoenix.util.PhoenixRuntime;
+import org.apache.phoenix.util.SchemaUtil;
+
+import java.io.IOException;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class SchemaExtractionProcessor {
+
+    Map<String, String> defaultProps = new HashMap<>();
+    Map<String, String> definedProps = new HashMap<>();
+
+    private static final String CREATE_TABLE = "CREATE TABLE %s";
+    private static final String CREATE_INDEX = "CREATE %sINDEX %s ON %s";
+    private static final String CREATE_VIEW = "CREATE VIEW %s%s AS SELECT * FROM %s%s";
+
+    private PTable table;
+    private Configuration conf;
+    public SchemaExtractionProcessor(Configuration conf, String pSchemaName, String pTableName)
+            throws SQLException {
+        this.conf = conf;
+        this.table = getPTable(pSchemaName, pTableName);
+    }
+
+    public String process() throws Exception {
+        String ddl = null;
+        if(this.table.getType().equals(PTableType.TABLE)) {
+            ddl = extractCreateTableDDL(this.table);
+        } else if(this.table.getType().equals(PTableType.INDEX)) {
+            ddl = extractCreateIndexDDL(this.table);
+        } else if(this.table.getType().equals(PTableType.VIEW)) {
+            ddl = extractCreateViewDDL(this.table);
+        }
+        return ddl;
```

Comment:
Since the object's immutable, any reason not to cache the generated DDL and return the cached copy if process() is called more than once?